Lux/src/image.cpp

```diff
@@ -0,0 +1,129 @@
+// Template base class for rasterized data - used as base for image classes and vector fields
+// Handles functions common to all of these classes.
+// Supports linear interpolated sampling and mip-mapping - multiple resolutions for anti-aliasing
+
+#include "image.hpp"
+
+#define I image<T>
+#define TT template< class T >
+
+TT void I::mip_it() { // mip it good
+    if( mip_me ) {
+        if( !mipped ) {
+            // allocate mip map memory
+            mipped = true;
+        }
+        if( !mip_utd ) {
+            // calculate mip-maps
+            mip_utd = true;
+        }
+    }
+}
+
+TT void I::de_mip() {  
+    mipped = false;
+    mip_utd = false;
+    // deallocate all mip-maps
+}
+
+TT void I::reset() { 
+    base.clear();
+    set_dim( { 0, 0 } );
+    de_mip(); 
+}
+
+TT void I::use_mip( bool m ) {
+    mip_me = m;
+    if( mip_me ) mip_it();
+    else {} // free mip-map memory?
+}
+
+TT vec2i I::get_dim() { return dim; }
+
+TT void I::set_dim( const vec2i& dims ) {
+    dim = dims;
+    ipbounds.set( { 0, 0 }, dim );
+}
+
+TT bb2f I::get_bounds() { return bounds; }
+TT void I::set_bounds( const bounding_box< float, 2 >& bb ) { bounds = bb; }
+
+// returns true if images have same dimensions
+TT bool I::compare_dims( const I& img ) { return ( dim == img.dim ); } 
+
+TT void I::fill( const T& c ) {
+    std::fill( base.begin(), base.end(), c );
+    mip_it();
+}
+
+TT I& I::operator += ( I& rhs ) {
```

Comment:
shouldn't this be const I& rhs  ?

Reply:
If I do this I receive the following error messages from `make`:
```joyhughes@Joys-MacBook-Pro build % make
Consolidate compiler generated dependencies of target common
[ 11%] Building CXX object CMakeFiles/common.dir/src/frgb.cpp.o
[ 22%] Building CXX object CMakeFiles/common.dir/src/image.cpp.o
In file included from /Users/joyhughes/src/Jen/Lux/src/image.cpp:5:
In file included from /Users/joyhughes/src/Jen/Lux/src/image.hpp:8:
In file included from /Users/joyhughes/src/Jen/Lux/src/vect2.hpp:18:
In file included from /Users/joyhughes/src/Jen/Lux/src/linalg.h:59:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/array:116:
/Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/algorithm:1993:21: error: no matching function for call to object of type '(lambda at /Users/joyhughes/src/Jen/Lux/src/image.cpp:75:81)'
        *__result = __binary_op(*__first1, *__first2);
                    ^~~~~~~~~~~
/Users/joyhughes/src/Jen/Lux/src/image.cpp:75:12: note: in instantiation of function template specialization 'std::__1::transform<std::__1::__wrap_iter<linalg::vec<float, 3> *>, std::__1::__wrap_iter<const linalg::vec<float, 3> *>, std::__1::__wrap_iter<linalg::vec<float, 3> *>, (lambda at /Users/joyhughes/src/Jen/Lux/src/image.cpp:75:81)>' requested here
    std :: transform( base.begin(), base.end(), rhs.base.begin(), base.begin(), []( T &a, T &b) { return a - b; } );
           ^
/Users/joyhughes/src/Jen/Lux/src/image.cpp:128:16: note: in instantiation of member function 'image<linalg::vec<float, 3>>::operator-=' requested here
template class image< frgb >;
               ^
/Users/joyhughes/src/Jen/Lux/src/image.cpp:75:81: note: candidate function not viable: 2nd argument ('const linalg::vec<float, 3>') would lose const qualifier
    std :: transform( base.begin(), base.end(), rhs.base.begin(), base.begin(), []( T &a, T &b) { return a - b; } );
                                                                                ^
/Users/joyhughes/src/Jen/Lux/src/image.cpp:75:81: note: conversion candidate of type 'linalg::vec<float, 3> (*)(linalg::vec<float, 3> &, linalg::vec<float, 3> &)'
In file included from /Users/joyhughes/src/Jen/Lux/src/image.cpp:5:
In file included from /Users/joyhughes/src/Jen/Lux/src/image.hpp:8:
In file included from /Users/joyhughes/src/Jen/Lux/src/vect2.hpp:18:
In file included from /Users/joyhughes/src/Jen/Lux/src/linalg.h:59:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/array:116:
/Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/algorithm:1993:21: error: no matching function for call to object of type '(lambda at /Users/joyhughes/src/Jen/Lux/src/image.cpp:75:81)'
        *__result = __binary_op(*__first1, *__first2);
                    ^~~~~~~~~~~
/Users/joyhughes/src/Jen/Lux/src/image.cpp:75:12: note: in instantiation of function template specialization 'std::__1::transform<std::__1::__wrap_iter<linalg::vec<float, 2> *>, std::__1::__wrap_iter<const linalg::vec<float, 2> *>, std::__1::__wrap_iter<linalg::vec<float, 2> *>, (lambda at /Users/joyhughes/src/Jen/Lux/src/image.cpp:75:81)>' requested here
    std :: transform( base.begin(), base.end(), rhs.base.begin(), base.begin(), []( T &a, T &b) { return a - b; } );
           ^
/Users/joyhughes/src/Jen/Lux/src/image.cpp:129:16: note: in instantiation of member function 'image<linalg::vec<float, 2>>::operator-=' requested here
template class image< vec2f >;
               ^
/Users/joyhughes/src/Jen/Lux/src/image.cpp:75:81: note: candidate function not viable: 2nd argument ('const linalg::vec<float, 2>') would lose const qualifier
    std :: transform( base.begin(), base.end(), rhs.base.begin(), base.begin(), []( T &a, T &b) { return a - b; } );
                                                                                ^
/Users/joyhughes/src/Jen/Lux/src/image.cpp:75:81: note: conversion candidate of type 'linalg::vec<float, 2> (*)(linalg::vec<float, 2> &, linalg::vec<float, 2> &)'
2 errors generated.
make[2]: *** [CMakeFiles/common.dir/src/image.cpp.o] Error 1
make[1]: *** [CMakeFiles/common.dir/all] Error 2
make: *** [all] Error 2
```

Reply:
Yes, that is weird. I don't understand why I am getting errors if I make the image `const` - maybe something having to do with linalg.h. I have to make the vector and float versions `const` in order to do expressions like `b *= 1.5f` where beauty is an fimage.